File: libs/power-log-parser/src/lib/parsers/data-script-changed-parser.ts
import { CardType, GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class DataScriptChangedParser implements ActionParser {
	readonly ParserName = 'DataScriptChangedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			((node.Object as TagChange).Name === (GameTag.TAG_SCRIPT_DATA_NUM_1 as number) ||
				(node.Object as TagChange).Name === (GameTag.TAG_SCRIPT_DATA_NUM_2 as number))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity);
		if (entity == null) {
			return null;
		}

		if (this.StateFacade.IsBattlegrounds() && entity.GetCardType() !== (CardType.ENCHANTMENT as number)) {
			return null;
		}

		const initialData1 = entity.GetTag(GameTag.TAG_SCRIPT_DATA_NUM_1, 0);
		const initialData2 = entity.GetTag(GameTag.TAG_SCRIPT_DATA_NUM_2, 0);
		let newData1 = initialData1;
		let newData2 = initialData2;
		if (tagChange.Name === (GameTag.TAG_SCRIPT_DATA_NUM_1 as number)) {
			newData1 = tagChange.Value;
		}
		if (tagChange.Name === (GameTag.TAG_SCRIPT_DATA_NUM_2 as number)) {
			newData2 = tagChange.Value;
		}
		if (initialData1 === newData1 && initialData2 === newData2) {
			return null;
		}

		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'DATA_SCRIPT_CHANGED',
				GameEventHelper.CreateProvider('DATA_SCRIPT_CHANGED', cardId, controllerId, entity.Id, this.StateFacade, {
					DataNum1: newData1,
					DataNum2: newData2,
					Updates: [{ CardId: cardId, EntityId: entity.Id, ControllerId: controllerId, DataNum1: newData1, DataNum2: newData2 }],
				}),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/DataScriptChangedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class DataScriptChangedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public DataScriptChangedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && ((node.Object as TagChange).Name == (int)GameTag.TAG_SCRIPT_DATA_NUM_1 || (node.Object as TagChange).Name == (int)GameTag.TAG_SCRIPT_DATA_NUM_2);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);
            if (entity == null)
            {
                return null;
            }

            // In BG, we need this only for enchantments
            if (StateFacade.IsBattlegrounds() && entity.GetCardType() != (int)CardType.ENCHANTMENT) 
            {
                return null;
            }

            var initialData1 = entity.GetTag(GameTag.TAG_SCRIPT_DATA_NUM_1, 0);
            var initialData2 = entity.GetTag(GameTag.TAG_SCRIPT_DATA_NUM_2, 0);
            int newData1 = initialData1;
            int newData2 = initialData2;
            if (tagChange.Name == (int)GameTag.TAG_SCRIPT_DATA_NUM_1)
            {
                newData1 = tagChange.Value;
            }
            if (tagChange.Name == (int)GameTag.TAG_SCRIPT_DATA_NUM_2)
            {
                newData2 = tagChange.Value;
            }
            if (initialData1 == newData1 && initialData2 == newData2)
            {
                return null;
            }

            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                 "DATA_SCRIPT_CHANGED",
                GameEvent.CreateProvider(
                    "DATA_SCRIPT_CHANGED",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    new {
                        DataNum1 = newData1,
                        DataNum2 = newData2,
                        Updates = new List<dynamic>() {
                            new { CardId = cardId, EntityId = entity.Id, ControllerId = controllerId, DataNum1 = newData1, DataNum2 = newData2 }
                        },
                    }),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
